flang/lib/Lower/CallInterface.cpp

```diff
@@ -0,0 +1,455 @@
+  auto characteristic =
+      Fortran::evaluate::characteristics::Procedure::Characterize(
+          procRef.proc(), foldingContext.intrinsics());
+  assert(characteristic && "Failed to get characteristic from procRef");
+  // The characteristic may not contain the argument characteristic if no
+  // the ProcedureDesignator has no interface.
+  if (!characteristic->HasExplicitInterface()) {
+    for (const auto &arg : procRef.arguments()) {
+      // Argument cannot be optional with implicit interface
+      const auto *expr = arg.value().UnwrapExpr();
+      assert(expr &&
+             "argument in call with implicit interface cannot be assumed type");
+      auto argCharacteristic =
+          Fortran::evaluate::characteristics::DummyArgument::FromActual(
+              "actual", *expr, foldingContext);
+      assert(argCharacteristic &&
+             "failed to characterize argument in implicit call");
+      characteristic->dummyArguments.emplace_back(
+          std::move(*argCharacteristic));
+    }
+  }
+  return *characteristic;
+}
+
+void Fortran::lower::CallerInterface::placeInput(
+    const PassedEntity &passedEntity, mlir::Value arg) {
+  assert(static_cast<int>(actualInputs.size()) > passedEntity.firArgument &&
+         passedEntity.firArgument >= 0 &&
+         passedEntity.passBy != CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = arg;
+}
+
+void Fortran::lower::CallerInterface::placeAddressAndLengthInput(
+    const PassedEntity &passedEntity, mlir::Value addr, mlir::Value len) {
+  assert(static_cast<int>(actualInputs.size()) > passedEntity.firArgument &&
+         static_cast<int>(actualInputs.size()) > passedEntity.firLength &&
+         passedEntity.firArgument >= 0 && passedEntity.firLength >= 0 &&
+         passedEntity.passBy == CallInterface::PassEntityBy::AddressAndLength &&
+         "bad arg position");
+  actualInputs[passedEntity.firArgument] = addr;
+  actualInputs[passedEntity.firLength] = len;
+}
+
+bool Fortran::lower::CallerInterface::verifyActualInputs() const {
+  if (getNumFIRArguments() != actualInputs.size())
+    return false;
+  for (auto arg : actualInputs) {
+    if (!arg)
+      return false;
+  }
+  return true;
+}
+
+template <typename T>
+static inline auto AsGenericExpr(T e) {
+  return Fortran::evaluate::AsGenericExpr(Fortran::common::Clone(e));
+}
+
+mlir::Value Fortran::lower::CallerInterface::getResultLength() {
+  // FIXME: technically, this is a specification expression,
+  // so it should be evaluated on entry of the region we are
+  // in, it can go wrong if the specification expression
+  // uses a symbol that may have change.
+  //
+  // The characteristic has to be explicit for such
+  // cases, so the allocation could also be handled on callee side, in such
+  // case. For now, protect with an unreachable.
+  assert(characteristic && "characteristic was not computed");
```

Comment:
It might be helpful to document somewhere that "characteristic" is the Fortran jargon for what other languages call function signatures and/or types. The bridge is a balancing act between the Fortran parser and the semantics expressed in MLIR. We should strive to make the code readable from both directions.

Reply:
I've added a comment to the top of the header:
> This utility uses the characteristic of Fortran procedures to operate, which
> is a term and concept used in Fortran to refer to the signature of a function
> or subroutine.